File: app/server/manager.py
# -*- coding: utf-8 -*-
from typing import Optional
import time
from PySide6.QtCore import QObject, Signal, QTimer
from app.models.project import Project, ServerState
from app.process.runner import ProcessRunner
from app.utils.ports import PortChecker, is_port_open, build_server_command
from app.utils.detection import extract_port_from_log


class ServerManager(QObject):
    state_changed = Signal(object)  # ServerState
    log_output = Signal(str)
    ready = Signal()
    port_detected = Signal(int, str)  # (port, url)
    port_mismatch = Signal(int, int, str)  # (configured_port, detected_port, active_url)

    def __init__(self, project: Project, parent=None):
        super().__init__(parent)
        self._project = project
        self._state = ServerState.STOPPED
        self._started_at: Optional[float] = None
        self._runner = ProcessRunner(self)
        self._port_checker: Optional[PortChecker] = None
        self._active_port: int = project.server.port
        self._active_url: str = project.server.url
        self._is_mismatch: bool = False
        self._port_was_already_occupied: bool = False
        
        self._runner.process_started.connect(self._on_runner_started)
        self._runner.output_ready.connect(self._on_runner_output)
        self._runner.error_ready.connect(self._on_runner_error)
        self._runner.process_finished.connect(self._on_runner_finished)
        self._runner.process_error.connect(self._on_runner_error)

    @property
    def state(self) -> ServerState:
        return self._state

    @property
    def project(self) -> Project:
        return self._project
        
    @project.setter
    def project(self, value: Project):
        self.update_project(value)

    @property
    def active_port(self) -> int:
        return self._active_port

    @property
    def active_url(self) -> str:
        return self._active_url

    @property
    def is_port_mismatch(self) -> bool:
        return self._is_mismatch

    @property
    def started_at(self) -> Optional[float]:
        """Epoch timestamp captured when the server entered RUNNING. None if stopped."""
        return self._started_at

    def start(self):
        if self._state not in (ServerState.STOPPED, ServerState.ERROR):
            return
            
        if not self._project.server.enabled:
            self.log_output.emit('Server not enabled')
            return
            
        self._set_state(ServerState.STARTING)
        self._active_port = self._project.server.port
        self._active_url = self._project.server.url
        self._is_mismatch = False
        
        configured_port = self._project.server.port
        host = self._project.server.url or '127.0.0.1'
        if '://' in host:
            host = host.split('://')[-1].split('/')[0].split(':')[0]

        # Check if configured port is already occupied before our server even starts
        if configured_port > 0 and is_port_open(host, configured_port):
            self._port_was_already_occupied = True
            self.log_output.emit(
                f"⚠️ [Port Notice] Configured port {configured_port} is already in use by another process. "
                f"The server may bind to a dynamic fallback port."
            )
        else:
            self._port_was_already_occupied = False
        
        raw_cmd = self._project.server.command
        cmd = build_server_command(raw_cmd, configured_port) if configured_port > 0 else raw_cmd
        cwd = self._project.path
        self.log_output.emit(f'Starting server: {cmd} in {cwd}')
        
        extra_env = {}
        if configured_port > 0:
            extra_env["PORT"] = str(configured_port)
            extra_env["VITE_PORT"] = str(configured_port)
            extra_env["SERVER_PORT"] = str(configured_port)

        self._runner.start(cmd, cwd, extra_env=extra_env)

    def stop(self):
        if self._state == ServerState.STOPPED:
            return
            
        self._set_state(ServerState.STOPPING)
        self._is_mismatch = False
        
        if self._port_checker:
            self._port_checker.stop()
            self._port_checker.deleteLater()
            self._port_checker = None
            
        self._runner.stop()
        self._started_at = None
        self._set_state(ServerState.STOPPED)

    def restart(self):
        if self._state in (ServerState.RUNNING, ServerState.STARTING, ServerState.ERROR):
            self.stop()
            QTimer.singleShot(500, self.start)
        else:
            self.start()

    def update_project(self, project: Project):
        self._project = project
        self._active_port = project.server.port
        self._active_url = project.server.url
        self._is_mismatch = False

    def _set_state(self, new_state: ServerState):
        if self._state != new_state:
            if new_state == ServerState.RUNNING and self._started_at is None:
                self._started_at = time.time()
            self._state = new_state
            self.state_changed.emit(new_state)

    def _enter_running(self):
        if self._started_at is None:
            self._started_at = time.time()
        self._set_state(ServerState.RUNNING)

    def _on_runner_started(self):
        port = self._project.server.port
        # If the port was already occupied before we started, we should not rely on PortChecker
        # for that port because it would immediately report ready for the WRONG process.
        if port > 0 and not self._port_was_already_occupied:
            host = self._project.server.url or '127.0.0.1'
            if '://' in host:
                host = host.split('://')[-1].split('/')[0].split(':')[0]
            
            timeout = self._project.server.startup_timeout or 30000
            self._port_checker = PortChecker(host, port, timeout, 250, self)
            self._port_checker.port_ready.connect(self._on_port_ready)
            self._port_checker.check_timeout.connect(self._on_port_timeout)
            self._port_checker.start()
        elif port <= 0:
            self._enter_running()
            self.ready.emit()

    def _on_port_ready(self):
        self._enter_running()
        self.ready.emit()
        self.log_output.emit(f'Server is ready on port {self._active_port}')

    def _on_port_timeout(self):
        self._set_state(ServerState.ERROR)
        self.log_output.emit(f'Server startup timeout after {self._project.server.startup_timeout}ms')

    def _on_runner_output(self, text: str):
        self.log_output.emit(text)

        # Dynamic port detection from live log output
        detected = extract_port_from_log(text)
        if detected:
            configured_port = self._project.server.port
            has_changed = (self._active_port != detected)
            is_different_from_config = (configured_port > 0 and detected != configured_port)

            self._active_port = detected
            self._active_url = f"http://localhost:{detected}"

            if is_different_from_config and not self._is_mismatch:
                self._is_mismatch = True
                self.port_mismatch.emit(configured_port, detected, self._active_url)
                self.log_output.emit(
                    f"⚠️ [PORT MISMATCH] Server started on port {detected}, differing from configured port {configured_port}"
                )
                self.log_output.emit(f"🔗 Active URL redirected to: {self._active_url}")
            elif has_changed:
                self.log_output.emit(f"[Auto-Detect] Server bound to active port {detected}")

            self.port_detected.emit(detected, self._active_url)

            if self._state == ServerState.STARTING:
                if self._port_checker:
                    self._port_checker.stop()
                    self._port_checker.deleteLater()
                    self._port_checker = None
                self._enter_running()
                self.ready.emit()

    def _on_runner_error(self, text: str):
        self.log_output.emit(f'[ERROR] {text}')

    def _on_runner_finished(self, exit_code: int, status: str):
        if self._port_checker:
            self._port_checker.stop()
            self._port_checker.deleteLater()
            self._port_checker = None

        self._started_at = None
        if exit_code == 0:
            self._set_state(ServerState.STOPPED)
        else:
            self._set_state(ServerState.ERROR)
            
        self.log_output.emit(f'Server process finished with exit code {exit_code} ({status})')

File: desktop-go/internal/server/manager.go
// Package server porta app/server/manager.py.
// El contexto stopCtx se cancela en Stop(): interrumpe la espera de puerto.
package server

import (
	"context"
	"fmt"
	"strconv"
	"strings"
	"sync"
	"time"

	"desktop-go/internal/models"
	"desktop-go/internal/process"
	"desktop-go/internal/utils/detection"
	"desktop-go/internal/utils/ports"
)

type Callbacks struct {
	OnStateChange  func(state models.ServerState)
	OnLog          func(line string, isError bool)
	OnReady        func()
	OnPortDetected func(port int, url string)
	OnPortMismatch func(configured, detected int, activeURL string)
}

type Manager struct {
	mu sync.Mutex

	project         models.Project
	state           models.ServerState
	startedAt       time.Time
	activePort      int
	activeURL       string
	isMismatch      bool
	portWasOccupied bool
	stopRequested   bool
	failureReason   string

	runner         *process.Runner
	cb             Callbacks

	stateListeners []func(models.ServerState)
	readyListeners []func()

	stopCtx      context.Context
	cancelWait   context.CancelFunc
	restartTimer *time.Timer

	// Inyección para tests. Defaults: ports.IsPortOpen / ports.WaitForPort.
	probePortFn func(host string, port int) bool
	waitPortFn  func(ctx context.Context, timeout time.Duration) error
}

// hostFromURL replica la extracción de host del Python:
// 'http://localhost:5173' → 'localhost'.
func hostFromURL(rawURL string) string {
	host := rawURL
	if host == "" {
		return "127.0.0.1"
	}
	if i := strings.Index(host, "://"); i >= 0 {
		host = host[i+3:]
	}
	if i := strings.IndexAny(host, "/:"); i >= 0 {
		host = host[:i]
	}
	if host == "" {
		return "127.0.0.1"
	}
	return host
}

func NewManager(project models.Project, cb Callbacks) *Manager {
	return &Manager{
		project:     project,
		state:       models.StateStopped,
		activePort:  project.Server.Port,
		activeURL:   project.Server.URL,
		cb:          cb,
		probePortFn: ports.IsPortOpen,
	}
}

func (m *Manager) State() models.ServerState {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.state
}

func (m *Manager) ActivePort() int {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.activePort
}

func (m *Manager) ActiveURL() string {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.activeURL
}

func (m *Manager) FailureReason() string {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.failureReason
}

// AddStateListener registra un oyente adicional de cambios de estado
// (usado por playwright.Manager para observar al servidor).
func (m *Manager) AddStateListener(fn func(models.ServerState)) {
	m.mu.Lock()
	m.stateListeners = append(m.stateListeners, fn)
	m.mu.Unlock()
}

// AddReadyListener registra un oyente adicional de servidor listo.
func (m *Manager) AddReadyListener(fn func()) {
	m.mu.Lock()
	m.readyListeners = append(m.readyListeners, fn)
	m.mu.Unlock()
}

// StartedAt replica started_at: timestamp al entrar RUNNING; false si parado.
func (m *Manager) StartedAt() (time.Time, bool) {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.startedAt, !m.startedAt.IsZero()
}

func (m *Manager) UpdateProject(p models.Project) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.project = p
	m.activePort = p.Server.Port
	m.activeURL = p.Server.URL
	m.isMismatch = false
}

func (m *Manager) log(line string, isError bool) {
	if m.cb.OnLog != nil {
		m.cb.OnLog(line, isError)
	}
}

// setState requiere m.mu NO tomado.
func (m *Manager) setState(s models.ServerState) {
	m.mu.Lock()
	if m.state == s {
		m.mu.Unlock()
		return
	}
	if s == models.StateRunning && m.startedAt.IsZero() {
		m.startedAt = time.Now()
	}
	m.state = s
	onChange := m.cb.OnStateChange
	listeners := make([]func(models.ServerState), len(m.stateListeners))
	copy(listeners, m.stateListeners)
	m.mu.Unlock()
	if onChange != nil {
		onChange(s)
	}
	for _, fn := range listeners {
		fn(s)
	}
}

// fireReady dispara OnReady y los readyListeners registrados (fuera del lock).
func (m *Manager) fireReady() {
	m.mu.Lock()
	onReady := m.cb.OnReady
	listeners := make([]func(), len(m.readyListeners))
	copy(listeners, m.readyListeners)
	m.mu.Unlock()
	if onReady != nil {
		onReady()
	}
	for _, fn := range listeners {
		fn()
	}
}

func (m *Manager) enterRunning() {
	m.mu.Lock()
	if m.startedAt.IsZero() {
		m.startedAt = time.Now()
	}
	m.mu.Unlock()
	m.setState(models.StateRunning)
}

// Start replica ServerManager.start().
func (m *Manager) Start() {
	m.mu.Lock()
	if m.state != models.StateStopped && m.state != models.StateError {
		m.mu.Unlock()
		return
	}
	if !m.project.Server.Enabled {
		m.mu.Unlock()
		m.log("Server not enabled", true)
		return
	}

	m.failureReason = ""
	m.activePort = m.project.Server.Port
	m.activeURL = m.project.Server.URL
	m.isMismatch = false

	configuredPort := m.project.Server.Port
	host := hostFromURL(m.project.Server.URL)
	command := m.project.Server.Command
	cwd := m.project.Path
	startupTimeoutMs := m.project.Server.StartupTimeout
	m.mu.Unlock()

	startupTimeout := time.Duration(startupTimeoutMs) * time.Millisecond
	if startupTimeout <= 0 {
		startupTimeout = 30 * time.Second
	}

	portWasOccupied := configuredPort > 0 && m.probePortFn(host, configuredPort)

	m.mu.Lock()
	m.portWasOccupied = portWasOccupied
	m.mu.Unlock()

	if portWasOccupied {
		m.log(fmt.Sprintf(
			"⚠️ [Port Notice] Configured port %d is already in use by another process. "+
				"The server may bind to a dynamic fallback port.", configuredPort), true)
	}

	cmd := command
	if configuredPort > 0 {
		cmd = ports.BuildServerCommand(command, configuredPort)
	}
	m.log(fmt.Sprintf("Starting server: %s in %s", cmd, cwd), false)

	extraEnv := map[string]string{}
	if configuredPort > 0 {
		extraEnv["PORT"] = strconv.Itoa(configuredPort)
		extraEnv["VITE_PORT"] = strconv.Itoa(configuredPort)
		extraEnv["SERVER_PORT"] = strconv.Itoa(configuredPort)
	}

	ctx, cancel := context.WithCancel(context.Background())

	m.mu.Lock()
	m.stopCtx = ctx
	m.cancelWait = cancel
	waitFn := m.waitPortFn
	m.mu.Unlock()

	runner := process.NewRunner(process.RunnerCallbacks{
		OnStdout:  func(line string) { m.onRunnerOutput(line, false) },
		OnStderr:  func(line string) { m.onRunnerOutput("[ERROR] "+line, true) },
		OnStarted: func() { m.onRunnerStarted(ctx, startupTimeout, waitFn) },
		OnFinished: func(exitCode int, status string) {
			cancel()
			m.onRunnerFinished(exitCode, status)
		},
		OnError: func(desc string) { m.log(desc, true) },
	})

	m.mu.Lock()
	m.runner = runner
	m.mu.Unlock()

	m.setState(models.StateStarting)
	if err := runner.Start(cmd, cwd, extraEnv); err != nil {
		cancel()
		m.mu.Lock()
		m.failureReason = "Failed to start: " + err.Error()
		m.mu.Unlock()
		m.setState(models.StateError)
	}
}

// Stop replica ServerManager.stop(): marca intención ANTES de matar para que
// taskkill /F no se reporte como crash.
func (m *Manager) Stop() {
	m.mu.Lock()
	if m.state == models.StateStopped {
		m.mu.Unlock()
		return
	}
	m.stopRequested = true
	m.failureReason = ""
	if m.restartTimer != nil {
		m.restartTimer.Stop()
		m.restartTimer = nil
	}
	cancel := m.cancelWait
	runner := m.runner
	m.mu.Unlock()

	if cancel != nil {
		cancel()
	}
	m.setState(models.StateStopping)
	if runner != nil {
		runner.Stop()
	}
	m.mu.Lock()
	m.startedAt = time.Time{}
	m.mu.Unlock()
	m.setState(models.StateStopped)
}

// Restart replica restart(): stop + arranque diferido 500ms si estaba vivo.
func (m *Manager) Restart() {
	m.mu.Lock()
	needsDelay := m.state == models.StateRunning ||
		m.state == models.StateStarting || m.state == models.StateError
	m.mu.Unlock()

	if needsDelay {
		m.Stop()
		m.mu.Lock()
		m.restartTimer = time.AfterFunc(500*time.Millisecond, m.Start)
		m.mu.Unlock()
	} else {
		m.Start()
	}
}

// onRunnerStarted replica _on_runner_started.
func (m *Manager) onRunnerStarted(ctx context.Context, startupTimeout time.Duration, waitFn func(context.Context, time.Duration) error) {
	m.mu.Lock()
	m.stopRequested = false // launch real: un exit anómalo vuelve a ser significativo
	port := m.project.Server.Port
	occupied := m.portWasOccupied
	url := m.project.Server.URL
	configuredTimeoutMs := m.project.Server.StartupTimeout
	m.mu.Unlock()

	if port > 0 && !occupied {
		host := hostFromURL(url)
		go func() {
			var err error
			if waitFn != nil {
				err = waitFn(ctx, startupTimeout)
			} else {
				err = ports.WaitForPort(ctx, host, port, startupTimeout, 250*time.Millisecond)
			}
			select {
			case <-ctx.Done():
				return // stop durante la espera: no tocar estado
			default:
			}
			if err == nil {
				m.onPortReady()
				return
			}
			if m.State() == models.StateStarting {
				m.mu.Lock()
				m.failureReason = fmt.Sprintf("Startup timeout after %dms", configuredTimeoutMs)
				m.mu.Unlock()
				m.setState(models.StateError)
				m.log(fmt.Sprintf("Server startup timeout after %dms", configuredTimeoutMs), true)
			}
		}()
	} else if port <= 0 {
		m.enterRunning()
		m.fireReady()
	}
}

func (m *Manager) onPortReady() {
	m.enterRunning()
	m.fireReady()
	port := m.ActivePort()
	m.log(fmt.Sprintf("Server is ready on port %d", port), false)
}

// onRunnerOutput replica _on_runner_output: log en vivo + detección dinámica
// de puerto + transición STARTING→RUNNING por log.
func (m *Manager) onRunnerOutput(line string, isError bool) {
	m.log(line, isError)

	m.mu.Lock()
	state := m.state
	m.mu.Unlock()
	if state != models.StateStarting && state != models.StateRunning {
		return
	}

	detected := detection.ExtractPortFromLog(line)
	if detected == 0 {
		return
	}

	var emitDetected bool
	var detectedURL string

	m.mu.Lock()
	configuredPort := m.project.Server.Port
	hasChanged := m.activePort != detected
	isDifferentFromConfig := configuredPort > 0 && detected != configuredPort

	m.activePort = detected
	m.activeURL = fmt.Sprintf("http://localhost:%d", detected)
	detectedURL = m.activeURL

	if isDifferentFromConfig && !m.isMismatch {
		m.isMismatch = true
		emitDetected = true
	} else if hasChanged || detected == configuredPort {
		emitDetected = true
	}
	mismatchNow := m.isMismatch && isDifferentFromConfig
	m.mu.Unlock()

	if mismatchNow {
		if m.cb.OnPortMismatch != nil {
			m.cb.OnPortMismatch(configuredPort, detected, detectedURL)
		}
		m.log(fmt.Sprintf("⚠️ [PORT MISMATCH] Server started on port %d, differing from configured port %d", detected, configuredPort), true)
		m.log(fmt.Sprintf("🔗 Active URL redirected to: %s", detectedURL), false)
	} else if hasChanged {
		m.log(fmt.Sprintf("[Auto-Detect] Server bound to active port %d", detected), false)
	}

	if emitDetected && m.cb.OnPortDetected != nil {
		m.cb.OnPortDetected(detected, detectedURL)
	}

	if m.State() == models.StateStarting {
		m.enterRunning()
		m.fireReady()
	}
}

// onRunnerFinished replica _on_runner_finished.
func (m *Manager) onRunnerFinished(exitCode int, status string) {
	m.mu.Lock()
	prevState := m.state
	stopReq := m.stopRequested
	m.startedAt = time.Time{}
	m.mu.Unlock()

	if stopReq || (status == "NormalExit" && exitCode == 0) {
		m.mu.Lock()
		m.stopRequested = false
		m.mu.Unlock()
		m.setState(models.StateStopped)
	} else {
		reason := "Server failed to start"
		if prevState == models.StateRunning {
			reason = "Server process crashed"
		}
		m.mu.Lock()
		m.failureReason = reason
		m.mu.Unlock()
		m.setState(models.StateError)
	}

	m.log(fmt.Sprintf("Server process finished with exit code %d (%s)", exitCode, status), status != "NormalExit")
}
